ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala

```diff
@@ -311,51 +367,138 @@ private[samplers] object BigSamplerAvro {
+      }
+    }
+  }
+  // scalastyle:on cyclomatic.complexity
+
+  //scalastyle:off method.length cyclomatic.complexity parameter.number
   private[samplers] def sampleAvro(sc: ScioContext,
-                         input: String,
-                         output: String,
-                         fields: Seq[String],
-                         samplePct: Float,
-                         seed: Option[Int]): Future[Tap[GenericRecord]] = {
+                                   input: String,
+                                   output: String,
+                                   fields: Seq[String],
+                                   fraction: Double,
+                                   seed: Option[Int],
+                                   distribution: Option[SampleDistribution],
+                                   distributionFields: Seq[String],
+                                   exact: Boolean)
+  : Future[Tap[GenericRecord]] = {
+    def buildKey(schema: Schema)(gr: GenericRecord): Set[String] = {
+      distributionFields.map(f => getAvroField(gr, f, schema).toString).toSet
+    }
+
     val schema = AvroIO.getAvroSchemaFromFile(input)
-    val outputParts =
-      if (output.endsWith("/")) output + "part*" else output + "/part*"
+    val outputParts = if (output.endsWith("/")) output + "part*" else output + "/part*"
     if (FileStorage(outputParts).isDone) {
       log.info(s"Reuse previous sample at $outputParts")
       Taps().avroFile(outputParts, schema = schema)
     } else {
       log.info(s"Will sample from: $input, output will be $output")
-      if (fields.isEmpty) {
-        val r = sc.avroFile[GenericRecord](input, schema)
-          .sample(false, samplePct)
-          .saveAsAvroFile(output, schema=schema)
-        sc.close().waitUntilDone()
-        r
-      } else {
-        val fieldsMissing = fields.filterNot(f => fieldInAvroSchema(schema, f))
-        if (fieldsMissing.nonEmpty) {
-          throw new NoSuchElementException(
-            s"""Could not locate field(s) ${fieldsMissing.mkString(",")} """ +
-              s"""in $input with schema $schema""")
-        }
-        val schemaSer = schema.toString(false)
-        @transient lazy val schemaSerDe = new Schema.Parser().parse(schemaSer)
-        val samplePct100 = samplePct * 100.0
-        val r = sc.avroFile[GenericRecord](input, schema)
-          .flatMap { e =>
-            val hasher = BigSampler.hashFun(seed=seed)
+      val schemaSer = schema.toString(false)
+      @transient lazy val schemaSerDe = new Schema.Parser().parse(schemaSer)
+      @transient lazy val logSerDe = LoggerFactory.getLogger(this.getClass)
+
+      val coll = sc.avroFile[GenericRecord](input, schema)
+
+      val sampledCollection: SCollection[GenericRecord] = (fields, distribution, exact) match {
+        case (Seq(), None, false) => coll.sample(withReplacement = false, fraction)
+
+        case (Seq(), Some(d), false) =>
```

Comment:
comments on which case does what would be helpful for the future

Reply:
👍